packages/scoped-elements/README.md

```diff
@@ -0,0 +1,276 @@
+
+## How it works
+
+`createScopedHtml` is going to define the custom element classes and creates a
+unique scoped tag for each of them if needed. In case those classes were
+previously defined then it's going to return the previous defined tag.
+
+Then, the `html` function provided by the `createScopedHtml` method is going to
+transform the template literal into another one replacing the tags used by the
+user by the ones defined by the custom elements. Finally, the transformed
+template literal is going to be processed by `lit-html`.
+
+`<my-button>${this.text}</my-button>` --> `<my-button-3>${this.text}</my-button-3>`
+
+## How to use it
+
+1. Import the `createScopedHtml` from `@open-wc/scoped-elements`.
+
+   ```js
+   import { createScopedHtml } from '@open-wc/scoped-elements';
+   ```
+
+2. Import the classes of the components you want to use.
+
+   ```js
+   import MyButton from './MyButton.js';
+   import MyPanel from './MyPanel.js';
+   ```
+
+3. Create the `html` function defining the tags you want to use for the
+   components.
+
+   ```js
+   const html = createScopedHtml({
+     'my-button': MyButton,
+     'my-panel': MyPanel,
+   });
+   ```
+
+4. Use your components in your html.
+
+   ```js
+   render() {
+     return html`
+       <my-panel class="panel">
+         <my-button>${this.text}</my-button>
+       </my-panel}>
+     `;
+   }
+   ```
+
+### Complete example
+
+```js
+import { css, LitElement } from 'lit-element';
+import { createScopedHtml } from '@open-wc/scoped-elements';
+import MyButton from './MyButton.js';
+import MyPanel from './MyPanel.js';
+
+const html = createScopedHtml({
+  'my-button': MyButton, // <-- bind the element with a tag in your html
+  'my-panel': MyPanel,
+});
+
+export default class MyElement extends LitElement {
+  static get styles() {
+    return css`
+      .panel {
+        padding: 10px;
+        background-color: grey;
+      }
+    `;
+  }
+
+  static get properties() {
+    return {
+      text: String,
+    };
+  }
+
+  render() {
+    return html`
+      <my-panel class="panel">
+        <my-button>${this.text}</my-button>
+      </my-panel}>
+    `;
+  }
+}
+```
+
+## Limitations
+
+1. Components imported via npm **MUST NOT** be self registering components
+   If a shared component (on npm) does not offer an export to the class alone, without the registration side effect, then this component can not be used. e.g. every component
+   that calls `customElement.define`
+
+   ```js
+   export class MyEl { ... }
+   customElement.define('my-el', MyEl);
+   ```
+
+   or uses the typescript decorator `customElement`
+
+   ```ts
+   @customElement('my-el')
+   export class MyEl { ... }
+   ```
+
+   only side effects free class exports may be used
+
+   ```js
+   export class MyEl { ... }
+   ```
+
+   Note: not share or components withing the main app shell are exempted from this rule.
+
+2. Every (!!!) component with sub component needs to use `scoped-elements`
```

Comment:
As this is an important limitation, it might be useful to add a reference to the use case?  
Specifically, Team Black should implement or should have implemented `scoped-elements` for both version 1 as version 2 of Feature A when using custom elements internally. (as well as any other feature that team exposes, ie Feature B)

Also, not sure if it is a separate limitation, but Feature A and Feature B should not use any global dependencies, only real ES modules.  
ie. if Feature A uses `DOMPurify`, and updates to a breaking version of `DOMPurify` in Feature A v2, it wil be breaking, as `DOMPurify` is installed in the global namespace. (module wrappers might not always do the trick)

Reply:
Global Scope is not handled at all with this this is only about registration of customElements under a uniq tag maybe uniqElement Tag is even a better name thats how i called it in my framework

Reply:
You're right.  😃 
But unfortunately, it's not only about the registration of customElements. As explained in the documentation, sub node_modules will be used to allow multiple versions. This means that if the breaking change (of Feature A) is a breaking change in a sub dependency (eg.`DOMPurify`), now, all of a sudden, both versions of `DOMPurify` will also be (tried to be) loaded in the same way. This will often lead to unexpected issues (global variables being overriden, global events being hit multiple times, ...), where before the loading of the global variable was only done once.  
Now, as long as all dependencies are real ES modules with no outside dependencies, it *should* not have unexpected consequences. So, this is more about a consequence of using multiple versions.

Reply:
added a Limitation: "Imported elements need to be fully side effect free (not only element registration)"